### PR TITLE
exporter/datadog: Translates otel exceptions to DataDog errors

### PR DESCRIPTION
**Pull request by mothershipper**
When an otel exception event is logged on a span, the datadog exporter will now translate the last exception event into an error on the datadog span.  If the semconv fields aren't added, or an exception event isn't logged, this code should fall back to the previous behavior.

It should also be noted that the golang library seems to deviate from the spec as detailed [here](https://github.com/open-telemetry/opentelemetry-go/issues/1491). But for libraries that conform to the spec, such as otel-JS, this should improve what the end user sees in DataDog.

**Description:** <Describe what has changed.>
The otel collector used to log all errors to datadog like so:

```javascript
error: { 
  message: "ERR_CODE_2"
}
```

When using the [`span.recordException`](https://github.com/open-telemetry/opentelemetry-specification/blob/master/specification/trace/api.md#record-exception) method, we should get much richer details about the error that occurred:

```javascript
error: { 
  message: "actual error message",
  stack: "something happened at line 67..."
  type: "HttpError"
}
```

**Testing:** 

Added a test to show that the semconv fields, when set on the last exception in a given span, will get translated to DataDog's `ext.Error*` fields. The existing tests should prove that the previous behavior still works in the absence of the exception event.

**Documentation:**

No public-facing changes, this should be a backwards-compatible upgrade.

### CODE REVIEW

**Review by ericmustin — APPROVED** [review]
good catch @mothershipper , appreciate the contribution here as I believe we've been off-spec around  errors before this.

Fwiw, I think we're planning to add more support for span events in the next sprint here. Would be keen to get your input on if your org is leveraging span events in any way and how you'd prefer to see them in the ddog ui.

**Comment by mothershipper** [comment]
@ericmustin We're still pretty early on our path towards tracing adoption, i.e. we've baked otel into our shared service libraries and are working on tweaking the collector to play nicely with DD. We haven't declared it generally available for the rest of the dev team to use, so our use-cases are still fairly conservative. At this point we're only using events to record exceptions, and it's more of an implementation detail than a feature.

That said, DD is our only trace visualization tool, so we probably wouldn't dive into events until they're supported. We'd be happy to be a sounding board and/or beta test any UI changes for them :)

File: exporter/datadogexporter/translate_traces.go [hunk 1]
```diff
@@ -465,14 +465,48 @@ func getSpanErrorAndSetTags(s pdata.Span, tags map[string]string) int32 {
...
+		// exception that escaped the scope of the span.
+		//
+		// TODO:
+		//  Seems that the spec has an attribute that hasn't made it to the collector yet -- "exception.escaped".
+		//  This seems optional (SHOULD vs. MUST be set), but it's likely that we want to bubble up the exception
+		//  that escaped the scope of the span ("exception.escaped" == true) instead of the last exception event
+		//  in the case that these events differ.
+		//
+		//  https://github.com/open-telemetry/opentelemetry-specification/blob/master/specification/trace/semantic_conventions/exceptions.md#attributes
+		attribs := pdata.NewAttributeMap()
```
Comment (anuraaga) [inline exporter/datadogexporter/translate_traces.go:481]: I believe `nil` would work ok even if an error isn't found, to avoid this unused allocation

File: exporter/datadogexporter/translate_traces_test.go [hunk 2]
```diff
@@ -282,6 +284,78 @@ func TestTracesTranslationErrorsAndResource(t *testing.T) {
...
+	span.Status().SetCode(pdata.StatusCodeError)
+
+	events := span.Events()
+	events.Resize(4)
+
+	events.At(0).SetTimestamp(pdataStartTime)
+	events.At(0).SetName("start")
+
+	events.At(1).SetTimestamp(pdataEndTime)
+	events.At(1).SetName(conventions.AttributeExceptionEventName)
```
Comment (anuraaga) [inline exporter/datadogexporter/translate_traces_test.go:326]: If possible, would recomend having test cases where the exception is first or last, to find off-by-one errors.

**Review by anuraaga — APPROVED** [review]
Small comments, thanks!

File: exporter/datadogexporter/translate_traces.go (hunk 1 above)
Comment (mothershipper) [reply exporter/datadogexporter/translate_traces.go:481]: Not sure I understand this suggestion -- AttributeMap zero-value does not work, nor does a nil pointer as it's child member (it dereferences in the `Get` call).

Snippet from the pdata package:

```go
// NewAttributeMap creates a AttributeMap with 0 elements.
func NewAttributeMap() AttributeMap {
	orig := []otlpcommon.KeyValue(nil)
	return AttributeMap{&orig}
}

func (am AttributeMap) Get(key string) (AttributeValue, bool) {
	for i := range *am.orig {
		akv := &(*am.orig)[i]
		if akv.Key == key {
			return AttributeValue{&akv.Value}, true
		}
	}
	return AttributeValue{nil}, false
}

```

I debated using nil on my first pass, but the resulting code wasn't as clean.  I don't mind refactoring it if we're worried about the performance hit.

File: exporter/datadogexporter/translate_traces_test.go (hunk 2 above)
Comment (mothershipper) [reply exporter/datadogexporter/translate_traces_test.go:326]: 👍

File: exporter/datadogexporter/translate_traces.go (hunk 1 above)
Comment (mothershipper) [reply exporter/datadogexporter/translate_traces.go]: Found a way to avoid using nil at all -- ended up not as gross as my first pass. Still not a huge fan of the triple nested `if`, but I can live with it 😄

Comment (anuraaga) [reply exporter/datadogexporter/translate_traces.go]: Thanks - sorry yeah forgot it's not a native map

**Comment by tonglil** [comment]
I want to check, is this working? Using version 0.22.0.
I am not able to get error "SpanEvents" to show up with any level of detail in Datadog.

They show up as:
image

The events are not translated into errors, but instead kept in the `event` tag key:
image


My code:
```go
func joinRPC(ctx context.Context, name string) (string, error) {
	ctx, span := trace.SpanFromContext(ctx).Tracer().Start(ctx, "joinRPC")
	defer span.End()

	v := url.Values{}
	v.Set("name", name)
	url := "http://localhost:8081/join?" + v.Encode()
	req, err := http.NewRequestWithContext(ctx, "GET", url, nil)
	if err != nil {
		return "", err
	}

	// Do call with instrumented http client
	resp, err := Do(req)
	if err != nil {
		return "", err
	}

	return string(resp), nil
}
```
```go
func Do(req *http.Request) ([]byte, error) {
	span := trace.SpanFromContext(req.Context())
	defer span.End()

	client := http.DefaultClient
	client.Transport = otelhttp.NewTransport(
		http.DefaultTransport,
		otelhttp.WithSpanOptions(trace.WithAttributes(
			// The http.route tag is necessary in order for Datadog to properly name spans with http.* tags (web type)
			// This is based on this span.type tag, or inferred https://docs.datadoghq.com/tracing/visualization/services_list/#services-types
			semconv.HTTPRouteKey.String(req.URL.Path),
		)),
		otelhttp.WithMessageEvents(otelhttp.ReadEvents, otelhttp.WriteEvents),
	)

	resp, err := client.Do(req)
	if err != nil {
		span.RecordError(err)
		return nil, err
	}
	defer resp.Body.Close()

	body, err := ioutil.ReadAll(resp.Body)
	if err != nil {
		span.RecordError(err)
		return nil, err
	}

	if resp.StatusCode != 200 {
		err := fmt.Errorf("StatusCode: %d, Body: %s", resp.StatusCode, body)
		span.RecordError(err)
		return nil, err
	}

	return body, nil
}
```

The collector's logging exporter reports errors are collected as spanevents:
```
2021-03-02T06:55:14.205Z	INFO	loggingexporter/logging_exporter.go:313	TracesExporter	{"#spans": 4}
2021-03-02T06:55:14.206Z	DEBUG	loggingexporter/logging_exporter.go:352	ResourceSpans #0
Resource labels:
     -> deployment.environment: STRING(local)
     -> service.instance.id: STRING(ExampleInstanceID)
     -> service.name: STRING(ExampleService)
     -> service.namespace: STRING(ExampleNamespace)
     -> service.version: STRING(10)
InstrumentationLibrarySpans #0
InstrumentationLibrary go.opentelemetry.io/contrib/instrumentation/net/http/otelhttp semver:0.17.0
Span #0
    Trace ID       : 6c8999ef614a272d3388b90939f779c1
    Parent ID      : 33913d993e4df58e
    ID             : dc2111738b7b6424
    Name           : doesn't matter
    Kind           : SPAN_KIND_CLIENT
    Start time     : 2021-03-02 06:55:07.884204 +0000 UTC
    End time       : 2021-03-02 06:55:07.885091148 +0000 UTC
    Status code    : STATUS_CODE_ERROR
    Status message : 
Attributes:
     -> http.route: STRING(/join)
     -> net.peer.ip: STRING()
     -> http.method: STRING(GET)
     -> http.url: STRING(http://localhost:8081/join?name=astraean)
     -> http.scheme: STRING(http)
     -> http.host: STRING(localhost:8081)
     -> http.flavor: STRING(1.1)
Events:
SpanEvent #0
     -> Name: error
     -> Timestamp: 1614668107885090000
     -> DroppedAttributesCount: 0
     -> Attributes:
         -> error.type: STRING(*net.OpError)
         -> error.message: STRING(dial tcp [::1]:8081: connect: connection refused)
Span #1
    Trace ID       : 6c8999ef614a272d3388b90939f779c1
    Parent ID      : 1afe57b3c8824d50
    ID             : 33913d993e4df58e
    Name           : joinRPC
    Kind           : SPAN_KIND_INTERNAL
    Start time     : 2021-03-02 06:55:07.884099 +0000 UTC
    End time       : 2021-03-02 06:55:07.885138636 +0000 UTC
    Status code    : STATUS_CODE_ERROR
    Status message : 
Events:
SpanEvent #0
     -> Name: error
     -> Timestamp: 1614668107885138000
     -> DroppedAttributesCount: 0
     -> Attributes:
         -> error.type: STRING(*url.Error)
         -> error.message: STRING(Get "http://localhost:8081/join?name=astraean": dial tcp [::1]:8081: connect: connection refused)
Span #2
    Trace ID       : 6c8999ef614a272d3388b90939f779c1
    Parent ID      : 1f08531b0b811930
    ID             : 1afe57b3c8824d50
    Name           : greetHandler
    Kind           : SPAN_KIND_INTERNAL
    Start time     : 2021-03-02 06:55:07.884063 +0000 UTC
    End time       : 2021-03-02 06:55:07.885168437 +0000 UTC
    Status code    : STATUS_CODE_OK
    Status message : 
Span #3
    Trace ID       : 6c8999ef614a272d3388b90939f779c1
    Parent ID      : 
    ID             : 1f08531b0b811930
    Name           : doesn't matter
    Kind           : SPAN_KIND_SERVER
    Start time     : 2021-03-02 06:55:07.884023 +0000 UTC
    End time       : 2021-03-02 06:55:07.885181046 +0000 UTC
    Status code    : STATUS_CODE_ERROR
    Status message : HTTP status code: 500
Attributes:
     -> net.transport: STRING(IP.TCP)
     -> net.peer.name: STRING([::1])
     -> net.peer.port: INT(53545)
     -> net.host.name: STRING(localhost)
     -> net.host.port: INT(8080)
     -> http.method: STRING(POST)
     -> http.target: STRING(/greet)
     -> http.server_name: STRING(server)
     -> http.user_agent: STRING(curl/7.64.1)
     -> http.request_content_length: INT(23)
     -> http.scheme: STRING(http)
     -> http.host: STRING(localhost:8080)
     -> http.flavor: STRING(1.1)
     -> http.route: STRING(/greet)
     -> http.read_bytes: INT(23)
     -> http.wrote_bytes: INT(22)
     -> http.status_code: INT(500)
Events:
SpanEvent #0
     -> Name: read
     -> Timestamp: 1614668107884088000
     -> DroppedAttributesCount: 0
     -> Attributes:
         -> http.read_bytes: INT(23)
SpanEvent #1
     -> Name: write
     -> Timestamp: 1614668107885168000
     -> DroppedAttributesCount: 0
     -> Attributes:
         -> http.wrote_bytes: INT(22)
```

I also have a stdout exporter, which is recording:
```json
[
	{
		"SpanContext": {
			"TraceID": "945024b9bfcb906d019d807ae2031e01",
			"SpanID": "a203e5997eb103fe",
			"TraceFlags": 1,
			"TraceState": null
		},
		"ParentSpanID": "42e0f65c60b3f4df",
		"SpanKind": 3,
		"Name": "doesn't matter",
		"StartTime": "2021-03-01T23:00:25.042886-08:00",
		"EndTime": "2021-03-01T23:00:25.043803983-08:00",
		"Attributes": [
			{
				"Key": "http.route",
				"Value": {
					"Type": "STRING",
					"Value": "/join"
				}
			},
			{
				"Key": "net.peer.ip",
				"Value": {
					"Type": "STRING",
					"Value": ""
				}
			},
			{
				"Key": "http.method",
				"Value": {
					"Type": "STRING",
					"Value": "GET"
				}
			},
			{
				"Key": "http.url",
				"Value": {
					"Type": "STRING",
					"Value": "http://localhost:8081/join?name=becurse"
				}
			},
			{
				"Key": "http.scheme",
				"Value": {
					"Type": "STRING",
					"Value": "http"
				}
			},
			{
				"Key": "http.host",
				"Value": {
					"Type": "STRING",
					"Value": "localhost:8081"
				}
			},
			{
				"Key": "http.flavor",
				"Value": {
					"Type": "STRING",
					"Value": "1.1"
				}
			}
		],
		"MessageEvents": [
			{
				"Name": "error",
				"Attributes": [
					{
						"Key": "error.type",
						"Value": {
							"Type": "STRING",
							"Value": "*net.OpError"
						}
					},
					{
						"Key": "error.message",
						"Value": {
							"Type": "STRING",
							"Value": "dial tcp [::1]:8081: connect: connection refused"
						}
					}
				],
				"Time": "2021-03-01T23:00:25.043803-08:00"
			}
		],
		"Links": null,
		"StatusCode": "Error",
		"StatusMessage": "",
		"HasRemoteParent": false,
		"DroppedAttributeCount": 0,
		"DroppedMessageEventCount": 0,
		"DroppedLinkCount": 0,
		"ChildSpanCount": 0,
		"Resource": [
			{
				"Key": "deployment.environment",
				"Value": {
					"Type": "STRING",
					"Value": "local"
				}
			},
			{
				"Key": "service.instance.id",
				"Value": {
					"Type": "STRING",
					"Value": "ExampleInstanceID"
				}
			},
			{
				"Key": "service.name",
				"Value": {
					"Type": "STRING",
					"Value": "ExampleService"
				}
			},
			{
				"Key": "service.namespace",
				"Value": {
					"Type": "STRING",
					"Value": "ExampleNamespace"
				}
			},
			{
				"Key": "service.version",
				"Value": {
					"Type": "STRING",
					"Value": "10"
				}
			}
		],
		"InstrumentationLibrary": {
			"Name": "go.opentelemetry.io/contrib/instrumentation/net/http/otelhttp",
			"Version": "semver:0.17.0"
		}
	},
	{
		"SpanContext": {
			"TraceID": "945024b9bfcb906d019d807ae2031e01",
			"SpanID": "42e0f65c60b3f4df",
			"TraceFlags": 1,
			"TraceState": null
		},
		"ParentSpanID": "b2361786a154df96",
		"SpanKind": 1,
		"Name": "joinRPC",
		"StartTime": "2021-03-01T23:00:25.042793-08:00",
		"EndTime": "2021-03-01T23:00:25.043835557-08:00",
		"Attributes": null,
		"MessageEvents": [
			{
				"Name": "error",
				"Attributes": [
					{
						"Key": "error.type",
						"Value": {
							"Type": "STRING",
							"Value": "*url.Error"
						}
					},
					{
						"Key": "error.message",
						"Value": {
							"Type": "STRING",
							"Value": "Get \"http://localhost:8081/join?name=becurse\": dial tcp [::1]:8081: connect: connection refused"
						}
					}
				],
				"Time": "2021-03-01T23:00:25.043836-08:00"
			}
		],
		"Links": null,
		"StatusCode": "Error",
		"StatusMessage": "",
		"HasRemoteParent": false,
		"DroppedAttributeCount": 0,
		"DroppedMessageEventCount": 0,
		"DroppedLinkCount": 0,
		"ChildSpanCount": 1,
		"Resource": [
			{
				"Key": "deployment.environment",
				"Value": {
					"Type": "STRING",
					"Value": "local"
				}
			},
			{
				"Key": "service.instance.id",
				"Value": {
					"Type": "STRING",
					"Value": "ExampleInstanceID"
				}
			},
			{
				"Key": "service.name",
				"Value": {
					"Type": "STRING",
					"Value": "ExampleService"
				}
			},
			{
				"Key": "service.namespace",
				"Value": {
					"Type": "STRING",
					"Value": "ExampleNamespace"
				}
			},
			{
				"Key": "service.version",
				"Value": {
					"Type": "STRING",
					"Value": "10"
				}
			}
		],
		"InstrumentationLibrary": {
			"Name": "go.opentelemetry.io/contrib/instrumentation/net/http/otelhttp",
			"Version": "semver:0.17.0"
		}
	},
	{
		"SpanContext": {
			"TraceID": "945024b9bfcb906d019d807ae2031e01",
			"SpanID": "b2361786a154df96",
			"TraceFlags": 1,
			"TraceState": null
		},
		"ParentSpanID": "4d5eea045a02e436",
		"SpanKind": 1,
		"Name": "greetHandler",
		"StartTime": "2021-03-01T23:00:25.042752-08:00",
		"EndTime": "2021-03-01T23:00:25.043862523-08:00",
		"Attributes": null,
		"MessageEvents": null,
		"Links": null,
		"StatusCode": "Unset",
		"StatusMessage": "",
		"HasRemoteParent": false,
		"DroppedAttributeCount": 0,
		"DroppedMessageEventCount": 0,
		"DroppedLinkCount": 0,
		"ChildSpanCount": 1,
		"Resource": [
			{
				"Key": "deployment.environment",
				"Value": {
					"Type": "STRING",
					"Value": "local"
				}
			},
			{
				"Key": "service.instance.id",
				"Value": {
					"Type": "STRING",
					"Value": "ExampleInstanceID"
				}
			},
			{
				"Key": "service.name",
				"Value": {
					"Type": "STRING",
					"Value": "ExampleService"
				}
			},
			{
				"Key": "service.namespace",
				"Value": {
					"Type": "STRING",
					"Value": "ExampleNamespace"
				}
			},
			{
				"Key": "service.version",
				"Value": {
					"Type": "STRING",
					"Value": "10"
				}
			}
		],
		"InstrumentationLibrary": {
			"Name": "go.opentelemetry.io/contrib/instrumentation/net/http/otelhttp",
			"Version": "semver:0.17.0"
		}
	},
	{
		"SpanContext": {
			"TraceID": "945024b9bfcb906d019d807ae2031e01",
			"SpanID": "4d5eea045a02e436",
			"TraceFlags": 1,
			"TraceState": null
		},
		"ParentSpanID": "0000000000000000",
		"SpanKind": 2,
		"Name": "doesn't matter",
		"StartTime": "2021-03-01T23:00:25.042724-08:00",
		"EndTime": "2021-03-01T23:00:25.043883683-08:00",
		"Attributes": [
			{
				"Key": "net.transport",
				"Value": {
					"Type": "STRING",
					"Value": "IP.TCP"
				}
			},
			{
				"Key": "net.peer.name",
				"Value": {
					"Type": "STRING",
					"Value": "[::1]"
				}
			},
			{
				"Key": "net.peer.port",
				"Value": {
					"Type": "INT64",
					"Value": 53916
				}
			},
			{
				"Key": "net.host.name",
				"Value": {
					"Type": "STRING",
					"Value": "localhost"
				}
			},
			{
				"Key": "net.host.port",
				"Value": {
					"Type": "INT64",
					"Value": 8080
				}
			},
			{
				"Key": "http.method",
				"Value": {
					"Type": "STRING",
					"Value": "POST"
				}
			},
			{
				"Key": "http.target",
				"Value": {
					"Type": "STRING",
					"Value": "/greet"
				}
			},
			{
				"Key": "http.server_name",
				"Value": {
					"Type": "STRING",
					"Value": "server"
				}
			},
			{
				"Key": "http.user_agent",
				"Value": {
					"Type": "STRING",
					"Value": "curl/7.64.1"
				}
			},
			{
				"Key": "http.request_content_length",
				"Value": {
					"Type": "INT64",
					"Value": 22
				}
			},
			{
				"Key": "http.scheme",
				"Value": {
					"Type": "STRING",
					"Value": "http"
				}
			},
			{
				"Key": "http.host",
				"Value": {
					"Type": "STRING",
					"Value": "localhost:8080"
				}
			},
			{
				"Key": "http.flavor",
				"Value": {
					"Type": "STRING",
					"Value": "1.1"
				}
			},
			{
				"Key": "http.route",
				"Value": {
					"Type": "STRING",
					"Value": "/greet"
				}
			},
			{
				"Key": "http.read_bytes",
				"Value": {
					"Type": "INT64",
					"Value": 22
				}
			},
			{
				"Key": "http.wrote_bytes",
				"Value": {
					"Type": "INT64",
					"Value": 22
				}
			},
			{
				"Key": "http.status_code",
				"Value": {
					"Type": "INT64",
					"Value": 500
				}
			}
		],
		"MessageEvents": null,
		"Links": null,
		"StatusCode": "Error",
		"StatusMessage": "HTTP status code: 500",
		"HasRemoteParent": false,
		"DroppedAttributeCount": 0,
		"DroppedMessageEventCount": 0,
		"DroppedLinkCount": 0,
		"ChildSpanCount": 1,
		"Resource": [
			{
				"Key": "deployment.environment",
				"Value": {
					"Type": "STRING",
					"Value": "local"
				}
			},
			{
				"Key": "service.instance.id",
				"Value": {
					"Type": "STRING",
					"Value": "ExampleInstanceID"
				}
			},
			{
				"Key": "service.name",
				"Value": {
					"Type": "STRING",
					"Value": "ExampleService"
				}
			},
			{
				"Key": "service.namespace",
				"Value": {
					"Type": "STRING",
					"Value": "ExampleNamespace"
				}
			},
			{
				"Key": "service.version",
				"Value": {
					"Type": "STRING",
					"Value": "10"
				}
			}
		],
		"InstrumentationLibrary": {
			"Name": "go.opentelemetry.io/contrib/instrumentation/net/http/otelhttp",
			"Version": "semver:0.17.0"
		}
	}
```

Any thoughts on what's missing? It seems like `recordError` (in my function or in the otelhttp library) is not translated into an error (not recording the `net.OpError` as the error.message) for Datadog?

I would like my errors to show up in Datadog like how the native and opentracing libraries show it:
image
image

**Comment by mothershipper** [comment]
@tonglil Golang is the one language not yet conforming to the spec:

https://github.com/open-telemetry/opentelemetry-go/issues/1491

The spec wants these attributes to be logged as `exception.*`, Golang was the only language to adopt `error.*` attributes. In the short term, you could call `AddEvent` yourself, `RecordError` is just a small wrapper over it.  But I'm going to try and wrap up the PR feedback this week, just not sure what the timeline is for them to merge / cut a new release.

TL;DR; this should work from Golang soon :)

**Comment by ericmustin** [comment]
@tonglil I think @mothershipper has nailed it here (far faster than i would've). @mothershipper if you find that this is blocked on the otel-go side, i can patch the datadogexporter to support both types. but would prefer the approach of fixing the off-spec client

**Comment by mothershipper** [comment]
@ericmustin might be blocked -- lots of back-and-forth on it, they're going to discuss on Thursday.  

I'm on the fence about patching the exporter. The obvious con is that "error" isn't a standard in the spec, and if it becomes part of the spec, it might be a conflicting definition. The upside of patching is that regardless of where otel-go lands on the issue, it wouldn't require apps to update their otel-go version to get nicer error messages in DD.

**Comment by ericmustin** [comment]
@mothershipper I have left comments in slack and in the issue where appropriate. If people feel like `error` events are a thing, it would be nice to have it added to the spec, but that won't preclude us (datadog) from supporting it before that if push comes to shove. the goal here, imo, should be user adoption, so if this ends up being a block i will try to accomodate

**Comment by ericmustin** [comment]
👋 Hey @tonglil , thanks to the work over in otel-go from @mothershipper @MrAlias , it looks like they've merged a fix here, see: https://github.com/open-telemetry/opentelemetry-go/pull/1492. I'd try upgrading to the next opentelemetry-go release when it's released, and seeing if that resolves the issue

**Comment by nauxliu** [comment]
The exception event is not translated to an error for me, I'm using the open-telemetry PHP SDK, I added a logging exporter for debugging, and there's the output, the event looks like conforms to the semantic convention.

```
Events:
SpanEvent #0
     -> Name: exception
     -> Timestamp: 2021-08-04 14:45:02.664243968 +0000 UTC
     -> DroppedAttributesCount: 0
     -> Attributes:
         -> exception.type: STRING(RuntimeException)
         -> exception.message: STRING(Test exception in trace)
         -> exception.stacktrace: STRING(#0 /app/vendor/laravel/.....)
```

**Comment by ericmustin** [comment]
@nauxliu that event lgtm, is the status being set to `error` though?

**Comment by nauxliu** [comment]
@ericmustin Thanks! that is the cause.
